Modules/fcntlmodule.c

```diff
@@ -173,79 +170,62 @@ fcntl_ioctl_impl(PyObject *module, int fd, unsigned int code,
 
     if (ob_arg != NULL) {
         if (PyArg_Parse(ob_arg, "w*:ioctl", &pstr)) {
-            char *arg;
             str = pstr.buf;
             len = pstr.len;
 
-            if (mutate_arg) {
-                if (len <= IOCTL_BUFSZ) {
-                    memcpy(buf, str, len);
-                    buf[len] = '\0';
-                    arg = buf;
-                }
-                else {
-                    arg = str;
-                }
-            }
-            else {
-                if (len > IOCTL_BUFSZ) {
-                    PyBuffer_Release(&pstr);
-                    PyErr_SetString(PyExc_ValueError,
-                        "ioctl string arg too long");
-                    return NULL;
-                }
-                else {
-                    memcpy(buf, str, len);
-                    buf[len] = '\0';
-                    arg = buf;
-                }
-            }
-            if (buf == arg) {
-                Py_BEGIN_ALLOW_THREADS /* think array.resize() */
-                ret = ioctl(fd, code, arg);
-                Py_END_ALLOW_THREADS
-            }
-            else {
-                ret = ioctl(fd, code, arg);
+            PyObject *o = PyBytes_FromStringAndSize(NULL, len);
+            if (o == NULL) {
+                return NULL;
             }
-            if (mutate_arg && (len <= IOCTL_BUFSZ)) {
+            char *buf = PyBytes_AS_STRING(o);
+
+            memcpy(buf, str, len);
+
+            Py_BEGIN_ALLOW_THREADS /* think array.resize() */
+            ret = ioctl(fd, code, buf);
+            Py_END_ALLOW_THREADS
+
+            if (mutate_arg) {
                 memcpy(str, buf, len);
             }
             PyBuffer_Release(&pstr); /* No further access to str below this point */
             if (ret < 0) {
+                Py_DECREF(o);
```

Comment:
`PyBuffer_Release()` and `Py_DECREF()` can modify `errno`. You should save its value after `ioctl()` call and restore it just before `PyErr_SetFromErrno()`.